kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,389 @@
+[kythe fact](https://kythe.io/docs/kythe-storage.html#_a_id_termfact_a_fact),
+repeated here:
+
+```
+analysis-type =  "/" | 1*path
+path          =  "/" word
+word          =  1*{LETTER|DIGIT|PUNCT}
+LETTER        =  // Unicode letter, category Lu, Ll, Lt, Lm, or Lo
+DIGIT         =  // Unicode digit, category Nd
+PUNCT         =  [-.@#$%&_+:()]
+```
+ 
+Analysis types whose leading segment is `/kythe` are reserved for use by
+Kythe. An example analysis type might be `/kythe/index/java`, indicating the
+Kythe indexer for Java.
+
+### Methods
+
+A driver defines the following JSON-RPC methods:
+
+-   `init(InitRequest) → InitReply`
+
+    The `init` method is a protocol handshake, and must be the first method
+    called by the analyzer upon startup. The request specifies the protocol
+    version understood by the analyzer. If the environment understands the
+    requested version, it echoes the version in its reply.
+
+    If any other method is called prior to a successful invocation of `init`, or
+    if the environment doesn't recognize the protocol version requested, the
+    environment will report ProtocolError and close the input channel.
+
+    This document defines the protocol `"kythe1"`.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":1,"method":"init","params":{"protocol":"kythe1", "output-encoding":"json"}}
+        « {"jsonrpc":"2.0","id":1,"result":{"protocol":"kythe1"}}
+
+-   `analyze(AnalyzeRequest) → AnalyzeReply`
+
+    Request a fresh compilation record from the environment, blocking until one
+    is available. Once an analysis has started, it is designated as **pending**
+    until the analyzer sends a `done` notification (see below). The driver MUST
+    report ProtocolError and close the input channel if the analyzer issues an
+    `analyze` request while another analysis is pending.
+    
+    The reply includes the input file(s) on which to begin analysis, as well as
+    the output file to which to emit analysis output. It reflects some of the
+    kythe.proto.CompilationUnit structure. The driver (and driver environment)
+    MUST make contents available for each file referenced in the compilation
+    record at the file system path specified in the compilation record.
+
+    The output file should contain the content appropriate to the
+    analysis-type, encoded as specified in the `InitRequest`. For Kythe index
+    analyses, this would be `kythe.common.Entry` objects. Other analyzers might
+    emit a "findings" structure.
+
+    Note that contents need not be literally copied. An overlay filesystem,
+    hardlinks, symlinks, or other may be used.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":2,"method":"analyze","params":{"types":["kythe/indexer/go"]}}
+        « {"jsonrpc":"2.0","id":2","result":{"input":"path/to/main.go", "output":"path/to/output.json"}}
+
+-   `vname(VNameRequest)`
+
+    Requests the complete VName to emit for the specified path and signature.
```

Comment:
Yes, it relieves the analyzer from having to thread through those two fields. More significantly, though, it opens up the possibility to handle the case when semantic entities "belong" to some other corpus and/or root: one could imagine wanting to associate files of dependencies with some "well known" corpus name, or that build-time generated files should have some specific root. In these cases, some component needs to "understand" the mapping of path to corpus/root, and it makes sense to insulate the myriad indexers from how to determine that mapping.

Reply:
The analyzer was already freed of the need to reason about vname mapping by the compilation record. As I understand it, the real change here is providing that information _dynamically_ during the analysis instead of ahead-of-time. That's actually an interesting change, and I'm not entirely clear on the motivation for it.

We talked about various ways to solve this issue in the original implementation, and settled on making the build system capture that knowledge in the CU (that's what the vname fields are for, in the CU and its required inputs). I'm not sure what the argument for requesting it on the fly is, but it's a viable approach. 

That said: We should probably discuss that question.

Reply:
Right, that was too early in the AM for a reply. The per-file VName still would be present in the CU proto. The real reason for this API is that with the files being read from the file system, the only remaining motivation to expose the raw compilation unit mapping (FileInput) to the analyzer would be to supply the VName overrides. Thus the introduction of this API to encapsulate things. While it does permit dynamic non-CU-supplied VName construction, that wasn't its primary intention.

Reply:
Ah! Yes, that makes sense.

Reply:
It also turns out that it's incredibly common to get this mapping wrong, particularly for generated files and (presumably) files from a different corpus.  Indexer authors routinely emit references using the raw "path" disregarding the CU/required_input VName, often up to hard-coding various values for "corpus" as well.  Abstracting this behind an interface should substantially help with this problem.

That said, I think the generic `vname()` interface may still be a bit too low-level.  There are a variety of other node kinds which have structural, factual or other VName requirements (tapp comes to mind), so it might be nice to express those requirements via an RPC/proto interface in some fashion.

But file/anchor VNames are the low-hanging fruit here.